contrib/endpoints/src/api_manager/check_security_rules.cc

```diff
@@ -14,92 +14,350 @@
+  }
+
+  Status status = Status::OK;
+  Status invalid = Status(Code::INTERNAL, kInvalidResponse);
+
+  const grpc_json *testResults = GetProperty(json, "testResults");
+  if (testResults == nullptr) {
+    env_->LogError("TestResults are null");
+    status = invalid;
+  } else {
+    const char *result = GetStringValue(testResults->child, "state");
+    if (result == nullptr) {
+      env_->LogInfo("Result state is empty");
+      status = invalid;
+    } else if (std::string(result) != "SUCCESS") {
+      status = Status(Code::PERMISSION_DENIED,
+                      std::string("Unauthorized ") +
+                          context->request()->GetRequestHTTPMethod() +
+                          " access to resource " +
+                          context->request()->GetRequestPath(),
+                      Status::AUTH);
+    }
+  }
+
+  grpc_json_destroy(json);
+  return status;
+}
+
+std::string AuthzChecker::GetOperation(std::string &httpMethod) {
+  if (httpMethod == "POST") {
+    return std::string("create");
+  }
+
+  if (httpMethod == "GET" || httpMethod == "HEAD" || httpMethod == "OPTIONS") {
+    return std::string("get");
+  }
+
+  if (httpMethod == "DELETE") {
+    return std::string("delete");
+  }
+
+  return std::string("update");
+}
+
+Status AuthzChecker::BuildTestRequestBody(
+    std::shared_ptr<context::RequestContext> context,
+    std::string &result_string) {
+  proto::TestRulesetRequest request;
+  proto::TestRulesetRequest::TestCase *test_case = request.add_test_cases();
+  std::string httpMethod = context->request()->GetRequestHTTPMethod();
+
+  test_case->set_service_name(context->service_context()->service_name());
+  test_case->set_resource_path(context->request()->GetRequestPath());
+  test_case->set_operation(AuthzChecker::GetOperation(httpMethod));
+  test_case->set_expectation(proto::TestRulesetRequest::TestCase::ALLOW);
+
+  ::google::protobuf::Value auth;
+  ::google::protobuf::Value token;
+  ::google::protobuf::Value claims;
+
+  Status status = utils::JsonToProto(context->auth_claims(), &claims);
+  if (!status.ok()) {
+    env_->LogError(std::string("Error creating Protobuf from claims") +
+                   status.ToString());
+    return status;
+  }
+
+  SetProtoValue(token, "token", claims);
+  SetProtoValue(auth, "auth", token);
+
+  Map<std::string, google::protobuf::Value> *variables =
+      test_case->mutable_variables();
+  (*variables)["request"] = auth;
+
+  status =
+      utils::ProtoToJson(request, &result_string, utils::JsonOptions::DEFAULT);
+  if (status.ok()) {
+    env_->LogDebug(std::string("PRotobuf to JSON string = ") + result_string);
+  } else {
+    env_->LogError(std::string("Error creating TestRulesetRequest") +
+                   status.ToString());
+  }
+
+  return status;
+}
+
+void AuthzChecker::SetProtoValue(::google::protobuf::Value &head,
+                                 std::string key,
+                                 ::google::protobuf::Value &value) {
```

Comment:
per style guide, the prototype should be
(const string& key, const Value& value, Value* head)